File: 012.jarakirefactor/recursividade/testerecursao6.erl
-module(testerecursao6).

-compile(export_all).

-import(loop, [for/3, while/2]).

x(V_m) ->
    st:get_new_stack(x),
    st:put({x, "m"}, {long, V_m}),
    st:put({x, "x"}, {long, 0}),
    case st:get(x, "m") =< 0 of
      true -> st:put({x, "x"}, {long, st:get(x, "m")});
      false ->
	  st:put({x, "x"},
		 {long,
		  x(st:get(x, "m") - 1) +
		    (x(st:get(x, "m") - 2) + x(st:get(x, "m") - 3))})
    end,
    Return = st:get(x, "x"),
    st:get_old_stack(x),
    Return.

main(V_args) ->
    st:new(),
    st:get_new_stack(main),
    st:put({main, "args"}, {'String', V_args}),
    st:put({main, "m"}, {int, 25}),
    st:put({main, "a"}, {long, x(st:get(main, "m"))}),
    io:format("~p~n", [st:get(main, "a")]),
    st:get_old_stack(main).

File: 012.jarakirefactor/djeiva/TesteRecursao6.java

public class TesteRecursao6 {
	public static long x(long m){
		long x;
		x = 0;
		if (m <= 0) 
			x = m; 
		else x = x(m-1) + x(m -2) + x(m-3);
		return x;
	}

	public static void main(String[] args) {
		int m;
		long a;
		m = 25;
		a = x(m);
		System.out.println(a);
	}
}
